File: scripts/gemini_rag/categories.py
# Category definitions for Concrete Diagnostician exam questions

CATEGORIES = {
    "materials": {
        "name": "コンクリート材料",
        "keywords": ["セメント", "骨材", "混和材", "混和剤", "水", "鉄筋"],
        "count": 37,  # 15 existing + 22 new
        "description": "コンクリートの構成材料に関する問題"
    },
    "mix-design": {
        "name": "配合設計",
        "keywords": ["水セメント比", "スランプ", "空気量", "単位水量", "単位セメント量",
                     "配合強度", "呼び強度", "細骨材率", "粗骨材", "ワーカビリティー",
                     "AE剤", "減水剤", "フライアッシュ", "高流動コンクリート"],
        "count": 25,  # 10 existing + 15 new
        "description": "コンクリートの配合設計に関する問題"
    },
    "construction": {
        "name": "施工",
        "keywords": ["打設", "締固め", "養生", "打継ぎ", "運搬", "型枠"],
        "count": 37,  # 15 existing + 22 new
        "description": "コンクリートの施工に関する問題"
    },
    "quality": {
        "name": "品質管理・検査",
        "keywords": ["圧縮強度試験", "非破壊検査", "コア採取", "反発硬度", "超音波"],
        "count": 38,  # 15 existing + 23 new
        "description": "品質管理と各種検査方法に関する問題"
    },
    "deterioration": {
        "name": "劣化・損傷",
        "keywords": ["中性化", "塩害", "ASR", "アルカリシリカ反応", "凍害", "化学的侵食", "ひび割れ"],
        "count": 50,  # 20 existing + 30 new
        "description": "コンクリートの劣化現象と損傷に関する問題"
    },
    "diagnosis": {
        "name": "調査・診断",
        "keywords": ["目視点検", "コア採取", "調査計画", "劣化予測", "健全度評価"],
        "count": 38,  # 15 existing + 23 new
        "description": "構造物の調査・診断方法に関する問題"
    },
    "repair": {
        "name": "補修・補強",
        "keywords": ["断面修復", "ひび割れ補修", "表面保護", "電気防食", "補強"],
        "count": 25,  # 10 existing + 15 new
        "description": "補修・補強工法に関する問題"
    }
}

def get_all_categories():
    """Get all category codes"""
    return list(CATEGORIES.keys())

def get_category_info(category_code: str):
    """Get category information by code"""
    return CATEGORIES.get(category_code)

def get_total_question_count():
    """Get total number of questions across all categories"""
    return sum(cat["count"] for cat in CATEGORIES.values())

File: constants/Categories.ts
/**
 * Category definitions and ordering for the concrete diagnostician app
 * This is the single source of truth for category order and labels
 */

// ===== FREEMIUM SETTINGS =====
// 各カテゴリの無料問題割合（40%）
export const FREE_QUESTION_RATIO = 0.4;

// カテゴリごとの総問題数
export const CATEGORY_QUESTION_COUNTS: Record<string, number> = {
  'materials': 36,
  'mix-design': 36,
  'construction': 36,
  'quality': 35,
  'deterioration': 36,
  'diagnosis': 36,
  'repair': 35,
};

/**
 * カテゴリ内の無料問題上限を取得
 */
export function getFreeQuestionLimit(category: string): number {
  const total = CATEGORY_QUESTION_COUNTS[category] || 0;
  return Math.floor(total * FREE_QUESTION_RATIO);
}

/**
 * 問題がロックされているかチェック（Pro会員でない場合）
 */
export function isQuestionLocked(category: string, questionNumber: number): boolean {
  const freeLimit = getFreeQuestionLimit(category);
  return questionNumber > freeLimit;
}

// ===== CATEGORY DEFINITIONS =====

// Canonical category order - used for consistent display across the app
export const CATEGORY_ORDER: string[] = [
  'materials',      // コンクリート材料
  'mix-design',     // 配合設計
  'construction',   // 施工
  'quality',        // 品質管理・検査
  'deterioration',  // 劣化・損傷
  'diagnosis',      // 調査・診断
  'repair',         // 補修・補強
];

// Japanese labels for each category
export const CATEGORY_LABELS: Record<string, string> = {
  'materials': 'コンクリート材料',
  'mix-design': '配合設計',
  'construction': '施工',
  'quality': '品質管理・検査',
  'deterioration': '劣化・損傷',
  'diagnosis': '調査・診断',
  'repair': '補修・補強',
};

/**
 * Sort an array of items by category using the canonical order
 * Items with unknown categories are placed at the end
 */
export function sortByCategory<T extends { category: string }>(items: T[]): T[] {
  return [...items].sort((a, b) => {
    const indexA = CATEGORY_ORDER.indexOf(a.category);
    const indexB = CATEGORY_ORDER.indexOf(b.category);
    // Unknown categories go to the end
    const orderA = indexA === -1 ? CATEGORY_ORDER.length : indexA;
    const orderB = indexB === -1 ? CATEGORY_ORDER.length : indexB;
    return orderA - orderB;
  });
}

/**
 * Get the Japanese label for a category code
 */
export function getCategoryLabel(category: string): string {
  return CATEGORY_LABELS[category] || category;
}
